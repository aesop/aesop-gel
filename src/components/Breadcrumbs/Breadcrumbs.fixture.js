const BreadcrumbsFixture = {
  classname: 'breadcrumbs',
  items: [
    {
      label: 'Suspendisse quis elit',
      id: 'suspendisse-quis-elit',
      url: '/',
    },
    { label: 'Maecenas luctus', id: 'maecenas-luctus', url: '/' },
  ],
  theme: 'dark',
};

export { BreadcrumbsFixture };
