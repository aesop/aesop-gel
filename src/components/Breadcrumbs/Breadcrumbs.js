import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { isObjectPopulatedArray } from '~/utils/objects';
import Hyperlink from '~/components/Hyperlink';
import styles from './Breadcrumbs.module.css';

const Breadcrumbs = ({ className, items, theme }) => {
  if (!items || !isObjectPopulatedArray(items)) {
    return null;
  }

  const classSet = cx(styles.base, styles[theme], className);

  return (
    <nav className={classSet}>
      <ul className={styles.list}>
        {items.map(item => (
          <li className={styles.item} key={item.id}>
            <Hyperlink
              className={styles.link}
              id={item.id}
              theme={theme}
              title={item.title}
              url={item.url}
            >
              {item.label}
            </Hyperlink>
          </li>
        ))}
      </ul>
    </nav>
  );
};

Breadcrumbs.propTypes = {
  className: PropTypes.string,
  items: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      id: PropTypes.string,
      url: PropTypes.string,
      title: PropTypes.string,
    }),
  ),
  theme: PropTypes.oneOf(['dark', 'light']),
};

Breadcrumbs.defaultProps = {
  className: undefined,
  items: undefined,
  theme: 'dark',
};

export default Breadcrumbs;
