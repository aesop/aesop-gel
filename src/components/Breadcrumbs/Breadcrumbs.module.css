@import 'styles/index';

.base {
  margin: 16px 0;
}

.list {
  padding: 0;
  margin: 0;
  list-style: 0;
}

.item {
  display: inline-block;
  padding: 0;

  &:not(:first-of-type)::before {
    display: inline-block;
    padding: 0 10px;
    content: '\2022';
    font-size: rem(20px);
    line-height: 1;
    transform: translateY(3px);
  }
}

.link {
  position: relative;
  display: inline-block;
  color: var(--color-dark-copy);
  font-size: rem(14px);

  &::after {
    display: block;
    border-bottom: solid 1px var(--color-dark-copy);
    margin-top: -3px;
    margin-top: 0;
    margin-bottom: -10px;
    content: '';
    transform: scaleX(0);
    transform-origin: left;
    transition: transform 250ms var(--easing-ease-out-cubic);
  }

  &:hover::after,
  &:focus::after {
    transform: scaleX(1);
  }
}
