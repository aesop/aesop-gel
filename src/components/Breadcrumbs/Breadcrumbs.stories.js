import React from 'react';
import { storiesOf } from '@storybook/react';
import { object, select } from '@storybook/addon-knobs';
import '~.storybook/storybook.module.css';
import Breadcrumbs from './Breadcrumbs';

storiesOf('Components/Breadcrumbs', module).add('Base component', () => (
  <Breadcrumbs
    items={object(
      'items',
      [
        {
          label: 'Suspendisse quis elit',
          id: 'suspendisse-quis-elit',
          url: '/',
        },
        { label: 'Maecenas luctus', id: 'maecenas-luctus', url: '/' },
      ],
      'Copy',
    )}
    theme={select('theme', ['dark', 'light'], 'dark', 'User Interface')}
  />
));
