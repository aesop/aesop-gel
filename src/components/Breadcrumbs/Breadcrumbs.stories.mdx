import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs } from '@storybook/addon-knobs';
import * as knobs from '@storybook/addon-knobs';
import Breadcrumbs from './Breadcrumbs';
import BreadcrumbsFixture from './Breadcrumbs.fixture';

<Meta
  title="Components/Breadcrumbs"
  component={Breadcrumbs}
  decorators={[withKnobs]}
/>

# Breadcrumbs

<Preview>
  <Story name="Base component">
    <Breadcrumbs
      items={knobs.object('items', BreadcrumbsFixture.items, 'Copy')}
      theme={knobs.select(
        'theme',
        ['dark', 'light'],
        BreadcrumbsFixture.theme,
        'User Interface',
      )}
    />
  </Story>
</Preview>

## API

<Props of={Breadcrumbs} />
