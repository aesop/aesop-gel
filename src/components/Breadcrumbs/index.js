/* istanbul ignore file */
import Breadcrumbs from './Breadcrumbs';

export default Breadcrumbs;
