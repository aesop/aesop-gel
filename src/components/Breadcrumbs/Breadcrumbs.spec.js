import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import BreadcrumbsFixture from './Breadcrumbs.fixture';
import Breadcrumbs from './Breadcrumbs';

describe('<Breadcrumbs />', () => {
  it('should be defined', () => {
    expect(Breadcrumbs).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(<Breadcrumbs items={BreadcrumbsFixture.items} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should return `null` if the length of the `items` prop is 0', () => {
    const { container } = render(<Breadcrumbs items={[]} />);
    expect(container.firstChild).toBeNull();
  });

  it('should render the corrct number of items', async () => {
    const { getAllByRole } = render(
      <Breadcrumbs items={BreadcrumbsFixture.items} />,
    );
    const listItems = getAllByRole('listitem');

    expect(listItems).toHaveLength(2);
  });
});
