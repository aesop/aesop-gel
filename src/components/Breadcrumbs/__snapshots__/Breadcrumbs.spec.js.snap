// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Breadcrumbs /> should render base component correctly 1`] = `
<div>
  <nav
    class="base dark"
    data-testid="data-testid-Breadcrumbs"
  >
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <a
          class="base left dark link"
          href="/"
          target="_self"
        >
          Suspendisse quis elit
        </a>
      </li>
      <li
        class="item"
      >
        <a
          class="base left dark link"
          href="/"
          target="_self"
        >
          Maecenas luctus
        </a>
      </li>
    </ul>
  </nav>
</div>
`;
